packages/babel-core/src/config/loading/files/configuration.js

```diff
@@ -113,17 +119,38 @@ function readIgnoreConfig(filepath) {
   };
 }
 
+const LOADING_CONFIGS = new Set();
 function readConfigJS(filepath) {
   if (!exists(filepath)) return null;
 
+  // The `require()` call below can make this code reentrant if a require hook like babel-register has been
+  // loaded into the system. That would cause Babel to attempt to compile the `.babelrc.js` file as it loads
+  // below. To cover this case, we auto-ignore re-entrant config processing.
+  if (LOADING_CONFIGS.has(filepath)) {
+    debug("Auto-ignoring usage of config %o.", filepath);
+    return {
+      filepath,
+      dirname: path.dirname(filepath),
+      options: {
+        // Return a pattern that matches all files. Any file depending on this config will be skipped when
+        // it is compiled, if the file is loaded from inside the '.babelrc.js' file.
+        ignore: [ /^/ ],
```

Comment:
I go back and forth if we should include this. The alternative is that we just return an empty object, so files that are referenced from inside `.babelrc.js` will act like the `.babelrc` file was empty, which would allow them to still be processed by presets passed directly to `babel-register` programmatically. That would allow _some_ transformation, but might just be _more_ confusing because then the files could end up _partially_ compiled.